pkg/filters/clusterpackages.go

```diff
@@ -191,92 +203,111 @@ func (f *ClusterPackagesFilter) fetchClusterResources(ctx context.Context, res *
-    }
-
-		repo, err = git.PlainCloneContext(ctx, repoDir, false, cloneOptions)
-		if err != nil {
-			return nil, errors.WrapPrefixf(err, "error cloning Git repository %s", pkg.Git.Repo)
+		var repo *git.Repository
+		if !isCached {
+			f.Logger.Debug().Msgf("Cloning repository %s to %s", pkg.Git.Repo, repoDir)
+
+			var auth ssh.AuthMethod
+
+			switch f.AuthMethod {
+			case AuthMethodKeyFile:
+				auth, err = ssh.NewPublicKeys("git", f.GitPrivateKey, "")
+				if err != nil {
+					return nil, errors.WrapPrefixf(err, "error retrieving Git private key information")
+				}
+
+			case AuthMethodSSHAgent:
+				if os.Getenv(AuthSockEnvVar) == "" {
+					return nil, errors.Errorf("Env variable %s must be defined to use ssh agent auth", AuthSockEnvVar)
+				}
+				auth, err = ssh.NewSSHAgentAuth("git")
+				if err != nil {
+					return nil, errors.WrapPrefixf(err, "error using ssh agent auth")
+				}
+
+			default:
+				repoUrl, err := url.Parse(pkg.Git.Repo)
+				if err != nil {
+					return nil, errors.WrapPrefixf(err, "failed to parse repo URL")
+				}
+
+				if repoUrl.Scheme != HTTPSScheme && repoUrl.Scheme != "" {
+					return nil, errors.Errorf("got invalid scheme %s for anonymous authentication, use https scheme instead", repoUrl.Scheme)
+				}
+				auth = nil
+			}
+
+			cloneOptions := &git.CloneOptions{
+				URL: pkg.Git.Repo,
+			}
+
+			if auth != nil {
+				cloneOptions.Auth = auth
+			}
+
+			repo, err = git.PlainCloneContext(ctx, repoDir, false, cloneOptions)
+			if err != nil {
+				return nil, errors.WrapPrefixf(err, "error cloning Git repository %s", pkg.Git.Repo)
+			}
+		} else {
+			f.Logger.Debug().Msgf("Using %s in %s", pkg.Git.Repo, repoDir)
+
+			repo, err = git.PlainOpen(repoDir)
+			if err != nil {
+				return nil, errors.WrapPrefixf(err, "error opening Git repository %s", pkg.Git.Repo)
+			}
 		}
-	} else {
-		f.Logger.Debug().Msgf("Using %s in %s", pkg.Git.Repo, repoDir)
 
-		repo, err = git.PlainOpen(repoDir)
+		w, err := repo.Worktree()
 		if err != nil {
-			return nil, errors.WrapPrefixf(err, "error opening Git repository %s", pkg.Git.Repo)
+			return nil, errors.WrapPrefixf(err, "error obtaining worktree for repository %s", pkg.Git.Repo)
 		}
-	}
 
-	w, err := repo.Worktree()
-	if err != nil {
-		return nil, errors.WrapPrefixf(err, "error obtaining worktree for repository %s", pkg.Git.Repo)
-	}
+		if err := w.Checkout(&git.CheckoutOptions{Hash: plumbing.NewHash(pkg.Git.Ref)}); err != nil {
+			return nil, errors.WrapPrefixf(err, "error checking out ref %s for repository %s", pkg.Git.Ref, pkg.Git.Repo)
+		}
 
-	if err := w.Checkout(&git.CheckoutOptions{Hash: plumbing.NewHash(pkg.Git.Ref)}); err != nil {
-		return nil, errors.WrapPrefixf(err, "error checking out ref %s for repository %s", pkg.Git.Ref, pkg.Git.Repo)
+		subDirectory = pkg.Git.Directory
 	}
 
 	reader := kio.LocalPackageReader{
-		PackagePath:    filepath.Join(repoDir, pkg.Git.Directory),
+		PackagePath:    filepath.Join(repoDir, subDirectory),
```

Comment:
In the case that the package is from a local directory, why do you need to copy it into the cache? Could you not just set `PackagePath` to the local package path? That way you wouldn't need the `copyFile` and `copyDir` functions?

Reply:
Yeah, that's probably a lot easier 👍